### PR TITLE
Added built in collection fixtures

### PR DESCRIPTION
**Pull request by naz**
refs https://github.com/TryGhost/Arch/issues/25

- The instance should have two built-in collections "index" and "featured". These have been filled through in-memory tricks before, now they should come pre-populated through fixtures mechanism.

---

### <samp>🤖 Generated by Copilot at cfdde0a</samp>

This pull request adds two built-in collections, index and featured, to the Ghost database schema and data. It uses a migration script and a fixtures file to create the collections and removes the redundant code from the collections service. The built-in collections allow users to easily access all posts or only featured posts.

### CODE REVIEW

File: ghost/core/core/server/data/migrations/versions/5.55/2023-07-07-14-15-32-add-built-in-collections.js
Comment (ErisDS) [inline ghost/core/core/server/data/migrations/versions/5.55/2023-07-07-14-15-32-add-built-in-collections.js:20]: Detail to polish: we should decide if this built-in should be called "Latest" rather than Index. I think given the aim to make this easy to understand for users - it probably should be.

Comment (naz) [reply ghost/core/core/server/data/migrations/versions/5.55/2023-07-07-14-15-32-add-built-in-collections.js:20]: Agreed, matching the product naming makes sense 👍 If we market it and use it as "Latest" in cards, it would be pretty confusing seeing "Index" in collections UI once we have one and figuring out what the hell this thing does 😄

Comment (naz) [reply ghost/core/core/server/data/migrations/versions/5.55/2023-07-07-14-15-32-add-built-in-collections.js:20]: One thing that will be confusing though is "latest" implies there's ordering applied to the posts in the collection, which we are not forcing explicitly yet. I do think we should make the default order of posts in collections more explicit though.

**Review by ErisDS — COMMENTED** [review]
I don't think we need to worry about sort order rn

File: ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js
Comment (naz) [inline ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js:14]: Hardcode to `0` - we don't need it yet.

File: ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js
Comment (naz) [reply ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js:55]: We should not be worrying about it. The whole concept of ordering should not have entered our codebase just yet as it only creates artificial problems like this one. The ordering will be 🔥 from this migration.

File: ghost/core/test/utils/fixtures/fixtures.json
Comment (naz) [inline ghost/core/test/utils/fixtures/fixtures.json:1138]: This notation was pretty counter intuitive to figure out... Basically we have to have a duplicate "key" and "value" here for [matching to work](https://github.com/TryGhost/Ghost/blob/57a851227cad6e30c400636daf3bed517078de75/ghost/core/core/server/data/schema/fixtures/FixtureManager.js#L43). And the `key = key === 0 && value ? value : key;` in the matcher makes the matching logic even more confusing as it kinda works for the first record if passed in as an array, but than stops working when more elements are added.

Comment (naz) [reply ghost/core/test/utils/fixtures/fixtures.json:1138]: Reading the tests somewhat helps to understand the logic. But initially it was pretty confusing getting this right as the expected intuitive syntax like:
```js
"latest": ["integrations", "grow", "sell", ....]
```
didn't work out of the box.

File: ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js
Comment (naz) [inline ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js:20]: This part might cause performance issues. Needs additional checks on an instance with large amount of posts.

**Comment by naz** [comment]
There are flaky tests coming from `ghost/core/test/e2e-api/admin/collections.test.js` - they fail 20% of a time when run on MySQL locally. Possible reason for the failure is a race condition in the event processing inside of the collections service.

File: ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js
Comment (naz) [reply ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js:20]: Ran the migration on a db with ~4K records. Took `94.457ms` on my local machine. Imo we should be fine here. @daniellockyer any extra checks you'd perform for this kind of batch insert?

Comment (naz) [reply ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js:20]: For the scalability target of `100,000` the napkin math is 94ms x 25 = `2350ms` to run this query on local machine. If we assume the Pro environment is 10x slower (interesting metric to actually find out maybe?) that would be ~23s or running time.

Comment (daniellockyer) [reply ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js:20]: FWIW you don't need to loop over chunks because that's what `batchInsert` does: https://knexjs.org/guide/utility.html#batchinsert

Comment (naz) [reply ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js:20]: TIL! Ty

File: ghost/collections/src/Collection.ts
Comment (allouis) [inline ghost/collections/src/Collection.ts:72]: Wondering if we should instead have:
```
if (this.slug === 'latest' || this.slug === 'featured') {
  throw new ValidationError('Cannot edit internal collections');
}
```

File: ghost/collections/src/CollectionsService.ts
Comment (allouis) [inline ghost/collections/src/CollectionsService.ts:405]: Should this be happening at the repository level? 🤔

File: ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js
Comment (allouis) [inline ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js:36]: Does this need a filter for `status: 'published'`?

File: ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js
Comment (allouis) [inline ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js:56]: Does this need a filter for `status: 'published'`?

**Review by allouis — APPROVED** [review]
Few small comments.

One thing though - I'm not sure about the use of an empty filter for the "latest" collection 🤔 What do you think about using a filter of ~`-id:null`~ `id:-null` or something like that, which would include all of the posts, but it allows our validation to be as simple as "All automatic collections have a filter"

File: ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js
Comment (naz) [reply ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js:36]: nope. we would store all the posts on the db layer and only filter by published on the application layer. the fixture for "latest" collection has no filter at the moment at all.

File: ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js
Comment (naz) [reply ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js:56]: same answer as above for "latest" collection.

File: ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js
Comment (allouis) [reply ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js:36]: Why do we do that? Is it going to make our pagination more difficult?

File: ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-15-55-add-built-in-collections.js
Comment (kevinansfield) [inline ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-15-55-add-built-in-collections.js:40]: Could we standardise on not having `Posts` in the collection title? So `Latest` and `Featured` rather than `Latest` and `Featured Posts` so that when used in the UI it's consistent?

**Review by allouis — CHANGES_REQUESTED** [review]
We need to update the collection name as Kevin has requested before merging

File: ghost/collections/src/CollectionsService.ts
Comment (naz) [reply ghost/collections/src/CollectionsService.ts:405]: true. ordering is respositorie's concert 👍

Comment (naz) [reply ghost/collections/src/CollectionsService.ts:405]: Note, when moved to the repository level there was a race condition happening in the post's event processing. So leaving this as is for current purposes, but will need to move this logic once the event processing is improved.

File: ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js
Comment (naz) [reply ghost/core/core/server/data/migrations/versions/5.55/2023-07-10-05-16-55-add-built-in-collection-posts.js:36]: It does, but keeps the separation of logic clean (we had a very brief chat about this [here](https://ghost.slack.com/archives/C05DK3SP801/p1688723868922009))

File: ghost/collections/src/Collection.ts
Comment (naz) [reply ghost/collections/src/Collection.ts:72]: We might 🤔 As for now we don't have a requirement for the built-in collections to be "immutable". Maybe something we'd add eventually if that comes up as a feature. It looks like an obvious one to have but don't think we need it just yet.

**Comment by naz** [comment]
> about the use of an empty filter for the "latest" collection 🤔 What do you think about using a filter of -id:null id:-null or something like that, which would include all of the posts, but it allows our validation to be as simple as "All automatic collections have a filter"

@allouis I feel like this is one of those that seems like a good idea but would add future questions like "why is this thing even here?" from anybody reading through the codebase. Pros: we would save an extra if statement(s). Cons: add a confusing statement, which would need extra explanation, which would not be visible in the DB schema but rather somewhere in comments inside of the collections service. 

I'd leave it as is. Unless there are more good reasons to go down this path.

**Review by allouis — APPROVED** [review]

**Review by daniellockyer — APPROVED** [review]

**Pull request merged**
